File: leetcode/Triangle.py
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        """
        Calculates the minimum path sum from top to bottom by modifying the
        triangle in-place, using a bottom-up dynamic programming approach.
        This approach has O(1) extra space complexity.
        """
        # Start from the second-to-last row and move upwards.
        for i in range(len(triangle) - 2, -1, -1):
            # For each element in the current row, update it with the minimum path sum
            # starting from that element.
            for j in range(len(triangle[i])):
                # The minimum path from triangle[i][j] is its own value plus the
                # minimum of the two adjacent elements in the row below.
                triangle[i][j] += min(triangle[i + 1][j], triangle[i + 1][j + 1])

        # After the loops, the top element of the triangle will contain the
        # overall minimum path sum from the original top to the bottom.
        return triangle[0][0]

File: leetcode/medium/Triangle.java
import java.util.List;

/**
 * Solves the Minimum Path Sum in a Triangle problem using Dynamic Programming.
 * The solution modifies the input triangle in-place to store intermediate minimum path sums,
 * effectively reducing the required extra space to O(1).
 */
class Solution {
    /**
     * Calculates the minimum path sum from the top to the bottom of the triangle.
     *
     * @param triangle A list of lists representing the triangle of numbers.
     * @return The minimum total path sum.
     */
    public int minimumTotal(List<List<Integer>> triangle) {
        // The algorithm uses Dynamic Programming, starting from the second-to-last row
        // and working its way up to the top. This is a bottom-up approach.

        // Start from the second-to-last row (index triangle.size() - 2)
        // because the last row (base) is the starting point for the sums.
        for (int i = triangle.size() - 2; i >= 0; i--) {
            // Iterate through each element in the current row 'i'.
            // Row 'i' has 'i + 1' elements.
            for (int j = 0; j <= i; j++) {
                // Get the current element's value.
                int currentVal = triangle.get(i).get(j);

                // Find the minimum path sum from the two adjacent elements in the row below (i+1).
                // The adjacent elements are at indices 'j' and 'j + 1'.
                int minBelow = Math.min(triangle.get(i + 1).get(j), triangle.get(i + 1).get(j + 1));

                // The new value for triangle.get(i).get(j) is the sum of its original value
                // and the minimum path sum found from its two children in the row below.
                // This updates the element in-place to store the minimum total sum
                // from that point down to the base.
                int newMinTotal = currentVal + minBelow;

                // Update the triangle with the new minimum total sum.
                triangle.get(i).set(j, newMinTotal);
            }
        }

        // After the loops complete, the top element (0, 0) will hold the
        // minimum total sum from the top to the bottom of the triangle.
        return triangle.get(0).get(0);
    }
}